### ISSUE TITLE
[BUG] Http4s routes do not compose as expected 

### ISSUE THREAD

**Issue by calvinlfer** [softwaremill/tapir#1387]
Tapir version: 0.19-M2

Scala version: 2.13.6

**Describe the bug**

**What is the problem?**
Individual HTTP4S routes generated from Tapir endpoints do not compose properly using the `<+>` which comes from HTTP4S but endpoints in a list compose properly using ` ZHttp4sServerInterpreter().from(List(endpoint1, endpoint2, ...).toRoutes`

it used to work before on 0.18 but doesn’t seem to work properly on 0.19.

**How to reproduce?**
Here is a small application that attempts to store and retrieve students.

```scala
package tapir.experiments

import cats.implicits._
import io.circe.generic.semiauto._
import org.http4s.HttpRoutes
import org.http4s.blaze.server.BlazeServerBuilder
import org.http4s.server.Router
import org.http4s.syntax.all._
import sttp.tapir.generic.auto._
import sttp.tapir.json.circe._
import sttp.tapir.server.http4s.ztapir._
import sttp.tapir.ztapir._
import zio._
import zio.blocking.Blocking
import zio.clock.Clock
import zio.interop.catz._

import scala.collection.SortedMap

object Example extends App {
  type Effect[+A] = RIO[Clock & Blocking, A]

  final case class StudentRequest(name: String, age: Int)
  object StudentRequest {
    implicit val studentRequestCodec: io.circe.Codec[StudentRequest] = deriveCodec[StudentRequest]
  }

  final case class Student(id: Int, name: String, age: Int)
  object Student {
    implicit val studentCodec: io.circe.Codec[Student] = deriveCodec[Student]
  }

  object StudentRoutes                                   {
    def make: UIO[HttpRoutes[Effect]] =
      Ref
        .make(SortedMap.empty[Int, Student])
        .map(new StudentRoutes(_))
        .map(_.routes)
  }
  class StudentRoutes(ref: Ref[SortedMap[Int, Student]]) {
    private val getAllEndpoint: ZServerEndpoint[Any, Unit, Unit, List[Student]] =
      endpoint.get
        .in("students")
        .out(jsonBody[List[Student]])
        .zServerLogic(_ => ref.get.map(_.values.toList))

    private val addStudentEndpoint: ZServerEndpoint[Any, StudentRequest, Unit, Student] =
      endpoint.post
        .in("students")
        .in(jsonBody[StudentRequest])
        .out(jsonBody[Student])
        .zServerLogic { req =>
          ref.modify { existing =>
            val lastId  = existing.lastOption.map(_._1).getOrElse(0)
            val id      = lastId + 1
            val student = Student(id, req.name, req.age)
            (student, existing + (id -> student))
          }
        }

    private val getAllRoute     = ZHttp4sServerInterpreter().from(getAllEndpoint).toRoutes
    private val addStudentRoute = ZHttp4sServerInterpreter().from(addStudentEndpoint).toRoutes

    val routes: HttpRoutes[Effect]    = getAllRoute <+> addStudentRoute
    val altRoutes: HttpRoutes[Effect] = ZHttp4sServerInterpreter().from(getAllEndpoint :: addStudentEndpoint :: Nil).toRoutes
  }

  override def run(args: List[String]): URIO[ZEnv, ExitCode] =
    StudentRoutes.make.flatMap { routes =>
      val httpApp = Router("/" -> routes).orNotFound

      ZIO.executor.flatMap(ec =>
        BlazeServerBuilder[RIO[Clock & Blocking, *]](ec.asEC)
          .bindHttp(port = 9001, host = "0.0.0.0")
          .withHttpApp(httpApp)
          .serve
          .compile
          .drain
      )
    }.exitCode
}
```

Using `routes` demonstrates the problem showing you can only get students via `GET /students` but you get an HTTP 405 when you try to store students via `POST /students` but `altRoutes` works correctly for both endpoints.

**Additional information**
```scala
name := "playground"

version := "0.1"

scalaVersion := "2.13.6"

scalacOptions ++= Seq(" -Xsource:3")

addCompilerPlugin("org.typelevel" % "kind-projector" % "0.13.0" cross CrossVersion.full)

libraryDependencies ++= {
  val tapir        = "com.softwaremill.sttp.tapir"
  val tapirVersion = "0.19.0-M2"
  Seq(
    "dev.zio"                       %% "zio"                            % "1.0.9",
    "dev.zio"                       %% "zio-interop-cats"               % "3.1.1.0",
    "io.circe"                      %% "circe-generic"                  % "0.14.1",
    "com.softwaremill.sttp.client3" %% "async-http-client-backend-cats" % "3.3.11",
    "org.tpolecat"                  %% "doobie-core"                    % "1.0.0-M5",
    "org.tpolecat"                  %% "doobie-h2"                      % "1.0.0-M5",
    "org.typelevel"                 %% "cats-effect"                    % "3.1.1",
    "ch.qos.logback"                 % "logback-classic"                % "1.2.4"
  ) ++ Seq(
    tapir %% "tapir-zio",
    tapir %% "tapir-zio-http4s-server",
    tapir %% "tapir-json-circe",
    tapir %% "tapir-swagger-ui-http4s",
    tapir %% "tapir-redoc-http4s",
    tapir %% "tapir-openapi-docs",
    tapir %% "tapir-openapi-circe-yaml"
  ).map(_                            % tapirVersion)
}
```

**Comment by calvinlfer** [comment]
As it seems, the issue is not exclusive to ZIO, there is a problem with Cats Effect as well
```scala
package tapir.experiments

import cats.effect._
import cats.implicits._
import io.circe.generic.semiauto._
import org.http4s.HttpRoutes
import org.http4s.blaze.server.BlazeServerBuilder
import org.http4s.server.Router
import org.http4s.syntax.all._
import sttp.tapir._
import sttp.tapir.generic.auto._
import sttp.tapir.json.circe._
import sttp.tapir.server.ServerEndpoint
import sttp.tapir.server.http4s.Http4sServerInterpreter

import scala.collection.SortedMap

object TapirCEExample extends IOApp {
  final case class StudentRequest(name: String, age: Int)
  object StudentRequest {
    implicit val studentRequestCodec: io.circe.Codec[StudentRequest] = deriveCodec[StudentRequest]
  }

  final case class Student(id: Int, name: String, age: Int)
  object Student {
    implicit val studentCodec: io.circe.Codec[Student] = deriveCodec[Student]
  }

  object StudentRoutes                                       {
    def make: IO[HttpRoutes[IO]] =
      Ref
        .of[IO, SortedMap[Int, Student]](SortedMap.empty[Int, Student])
        .map(new StudentRoutes(_))
        .map(_.altRoutes)
  }
  class StudentRoutes(ref: Ref[IO, SortedMap[Int, Student]]) {
    private val getAllEndpoint: ServerEndpoint[Unit, Unit, List[Student], Any, IO] =
      endpoint.get
        .in("students")
        .out(jsonBody[List[Student]])
        .serverLogic[IO](_ => ref.get.map(_.values.toList).map(Right(_)))

    private val addStudentEndpoint: ServerEndpoint[StudentRequest, Unit, Student, Any, IO] =
      endpoint.post
        .in("students")
        .in(jsonBody[StudentRequest])
        .out(jsonBody[Student])
        .serverLogic[IO] { req =>
          ref
            .modify { existing =>
              val lastId  = existing.lastOption.map(_._1).getOrElse(0)
              val id      = lastId + 1
              val student = Student(id, req.name, req.age)
              (existing + (id -> student), student)
            }
            .map(_.asRight)
        }

    private val getAllRoute     = Http4sServerInterpreter[IO]().toRoutes(getAllEndpoint)
    private val addStudentRoute = Http4sServerInterpreter[IO]().toRoutes(addStudentEndpoint)

    val routes: HttpRoutes[IO]    = getAllRoute <+> addStudentRoute
    val altRoutes: HttpRoutes[IO] = Http4sServerInterpreter[IO]().toRoutes(getAllEndpoint :: addStudentEndpoint :: Nil)
  }

  override def run(args: List[String]): IO[ExitCode] =
    StudentRoutes.make
      .flatMap { routes =>
        val httpApp = Router("/" -> routes).orNotFound

        IO.executionContext.flatMap(ec =>
          BlazeServerBuilder[IO](ec)
            .bindHttp(port = 9001, host = "0.0.0.0")
            .withHttpApp(httpApp)
            .serve
            .compile
            .drain
        )
      }
      .as(ExitCode.Error)
}
```

Next up I'll try to create HTTP4S routes directly and report back

**Comment by calvinlfer** [comment]
Creation of HTTP4S routes work as intended (0.23-RC1)
```scala
package tapir.experiments

import cats.effect._
import cats.implicits._
import io.circe.generic.semiauto.deriveCodec
import org.http4s.blaze.server.BlazeServerBuilder
import org.http4s.circe.CirceEntityCodec._
import org.http4s.dsl.Http4sDsl
import org.http4s.implicits._
import org.http4s.server.Router
import org.http4s.{DecodeFailure, HttpRoutes}
import io.circe
import fs2._

import scala.collection.SortedMap
import scala.concurrent.ExecutionContext

object Http4sExample extends IOApp {
  final case class StudentRequest(name: String, age: Int)

  object StudentRequest {
    implicit val studentRequestCodec: circe.Codec[StudentRequest] = deriveCodec[StudentRequest]
  }

  final case class Student(id: Int, name: String, age: Int)

  object Student {
    implicit val studentCodec: circe.Codec[Student] = deriveCodec[Student]
  }

  object StudentRoutes {
    def make[F[_]: Concurrent]: F[HttpRoutes[F]] =
      Ref
        .of[F, SortedMap[Int, Student]](SortedMap.empty[Int, Student])
        .map(new StudentRoutes[F](_))
        .map(_.routes)
  }
  class StudentRoutes[F[_]: Concurrent](ref: Ref[F, SortedMap[Int, Student]]) extends Http4sDsl[F] {
    val getAllRoute: HttpRoutes[F] = HttpRoutes.of[F] { case GET -> Root / "students" =>
      ref.get
        .map(_.values)
        .flatMap(Ok(_))
    }

    val insertStudentRoute: HttpRoutes[F] = HttpRoutes.of[F] { case req @ POST -> Root / "students" =>
      req
        .as[StudentRequest]
        .flatMap { req =>
          ref.modify { existing =>
            val lastId  = existing.lastOption.map(_._1).getOrElse(0)
            val id      = lastId + 1
            val student = Student(id, req.name, req.age)
            (existing + (id -> student), student)
          }
        }
        .flatMap(Created(_))
        .handleErrorWith { case failure: DecodeFailure => BadRequest(failure.getMessage) }
    }

    val routes: HttpRoutes[F] = getAllRoute <+> insertStudentRoute
  }

  def server[F[_]: Async](ec: ExecutionContext): Stream[F, ExitCode] =
    Stream
      .eval(StudentRoutes.make[F])
      .flatMap { routes =>
        val httpApp = Router[F]("/" -> routes).orNotFound
        BlazeServerBuilder[F](ec)
          .bindHttp(port = 9001, host = "0.0.0.0")
          .withHttpApp(httpApp)
          .serve
      }

  override def run(args: List[String]): IO[ExitCode] =
    Stream
      .eval(IO.executionContext)
      .flatMap(server[IO](_))
      .compile
      .lastOrError
}
```

I believe the problem lies with Tapir

**Comment by PawelJ-PL** [comment]
Hi,

I added a unit test that proves the existence of this bug

**Comment by calvinlfer** [comment]
Awesome, thank you @PawelJ-PL

**Comment by adamw** [comment]
In a way, it's a feature, not a bug ;)

The change comes from https://github.com/softwaremill/tapir/pull/1356, which causes requests where the path matched, but the method didn't, to return a `405` response instead of "no match", causing the host library to check the next endpoint. So when you interpret individual endpoints, you get a different response than if you interpret them as a group.

This can be configured by setting the `rejectInterceptor` in `Http4sServerOptions` to `None`. I think in general we'd like to move from interpreting individual endpoints to interpreting lists of endpoints - this probably needs more coverage in the docs / release notes, but maybe there are some downside to interpreting endpoints as lists?

**Comment by PawelJ-PL** [comment]
I keep endpoints (no server logic) and routes (mostly logic) separate. Therefore, providing endpoints as a single list would be difficult.

**Comment by adamw** [comment]
I think a good middle ground will be to disable the reject interceptor for the interpreter methods which accept individual endpoints

**Comment by PawelJ-PL** [comment]
Fair enough for me. I think it is worth making this clear in the documentation, because it is breaking change.

**Comment by calvinlfer** [comment]
Thanks for clarifying guys! It would be great to have this in the documentation as this was a breaking change for us as well

**Comment by adamw** [comment]
Fixed by #1388
